File: frontend/src/objects/contest/response/ContestSummary.ts
import type { ContestDescription } from '@/objects/contest/ContestDescription'
import type { ContestId } from '@/objects/contest/ContestId'
import type { ContestSlug } from '@/objects/contest/ContestSlug'
import type { ContestTitle } from '@/objects/contest/ContestTitle'
import type { AuditFields } from '@/objects/shared/AuditFields'
import type { ResourceAccessPolicy } from '@/objects/shared/access/ResourceAccessPolicy'
import type { UserIdentity } from '@/objects/user/UserIdentity'
import type { ContestRegistrationStatus } from '@/objects/contest/response/ContestRegistrationStatus'

/** 比赛摘要响应；用于列表展示并携带当前会话是否可查看详情。 */
export type ContestSummary = AuditFields & {
  id: ContestId
  slug: ContestSlug
  title: ContestTitle
  description: ContestDescription
  startAt: string
  endAt: string
  accessPolicy: ResourceAccessPolicy
  registrationStatus: ContestRegistrationStatus
  canViewDetail: boolean
  author: UserIdentity | null
}

File: backend/src/main/scala/domains/contest/objects/response/ContestSummary.scala
package domains.contest.objects.response

import domains.contest.objects.*
import domains.user.objects.UserIdentity
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import shared.objects.access.ResourceAccessPolicy

import java.time.Instant
import scala.util.Try

/** 比赛列表摘要响应，包含详情可见标记但不携带赛题列表。 */
final case class ContestSummary(
  id: ContestId,
  slug: ContestSlug,
  title: ContestTitle,
  description: ContestDescription,
  startAt: Instant,
  endAt: Instant,
  accessPolicy: ResourceAccessPolicy,
  registrationStatus: ContestRegistrationStatus,
  canViewDetail: Boolean,
  author: Option[UserIdentity],
  createdAt: Instant,
  updatedAt: Instant
)

/** 提供比赛摘要 JSON codec，并显式处理 Instant 字符串格式。 */
object ContestSummary:
  private given Encoder[Instant] = Encoder.encodeString.contramap(_.toString)
  private given Decoder[Instant] = Decoder.decodeString.emap { value =>
    Try(Instant.parse(value)).toEither.left.map(_.getMessage)
  }

  given Encoder[ContestSummary] = deriveEncoder[ContestSummary]
  given Decoder[ContestSummary] = deriveDecoder[ContestSummary]
